lib/devise/strategies/database_authenticatable.rb

```diff
@@ -24,6 +25,17 @@ def authenticate!
           Devise.paranoid ? fail(:invalid) : fail(:not_found_in_database)
         end
       end
+
+      private
+
+      def regenerate_bcrypt_password_hash(resource, password)
+        if !Devise.respond_to?(:encryptor) &&
+            ::BCrypt::Password.new(resource.encrypted_password).cost != resource.class.stretches
+          resource.password = password
+          resource.skip_password_change_notification!
+          resource.save
```

Comment:
Please pay attention here.
`update_column` should be used here, but sourcelevel-bot consider this as a mistake, but it isn't - https://github.com/plataformatec/devise/pull/5145#discussion_r328703342.

Reply:
@mainameiz what do you mean by "Please pay attention here."?

Thanks for referencing my own comment regarding the usage of `update_column`. I used it initially and then during a refactoring iteration found the usage of `resource.skip_password_change_notification!` in the repo which (as far as I remember) in conjunction with `resource.save` seemed like a more proper implementation. Again, as far as I remember that's the only callback we need to avoid during the update. I haven't looked into the related repo code again now, but what I just wrote should be trustworthy :) I'd appreciate if you can review and verify.

Reply:
I'm not really familiar with this codebase but I suppose that `resource` variable is a instance of `User` model of some application. So it's state in database may be not valid. And if it's not valid then `resource.save` will not update the password.
I suppose, Devise should regenerate the digest nevertheless, so `update_column` should be used here.

Reply:
Thanks for the feedback. Indeed, `resource.save` may fail if a `resource` is not valid.

I recalled why `#update_column` can't be used. See [the Travis CI build](https://travis-ci.org/plataformatec/devise/builds/590032290) for [the initial commit](https://github.com/plataformatec/devise/commit/496dfe9fe666707b190f4de42e1da19bc0d346ee) where `#update_column` was used. All jobs with `DEVISE_ORM=mongoid`  failed because it's an Active Record method.

Reply:
What about `resource.save(validate: false)`?

Reply:
From https://guides.rubyonrails.org/v6.0/active_record_callbacks.html#running-callbacks:  
>The following methods **trigger** callbacks:
...
save(validate: false)
...

On the contrary we want to avoid callbacks like sending the password change notification.

Reply:
@sergey-alekseev 
> On the contrary we want to avoid callbacks like sending the password change notification

Invoking `skip_password_change_notification!` as the name tells, skips the callback you mentioned.

Your original code `resource.save` also triggers callbacks, so I didn't get your point.

Reply:
Maybe this could solve it:
```suggestion
          update_column(:encrypted_password, BCrypt::Password.create(password, cost: resource.class.stretches).to_s)
```

Reply:
Ah, I thought you want to use  
```ruby
resource.save(validate: false)
```  
instead of  
```ruby
resource.skip_password_change_notification!
resource.save
```
because of the previous context in this thread.  

Now I see your suggestion was to replace `resource.save` with `resource.save(validate: false)`. This totally makes sense. `resource.save(validate: false)` is extensively used in the codebase (e.g. look into https://github.com/plataformatec/devise/pull/4194). I made the change in https://github.com/plataformatec/devise/pull/5145/commits/31998a959d339b5352ad42861df2b2dbc1aef8f7#diff-d1ac35f7ea071d6c944aeb7ee4cd06cfR36 and covered this change in tests in https://github.com/plataformatec/devise/pull/5145/commits/31998a959d339b5352ad42861df2b2dbc1aef8f7#diff-0cd3b7071e3422d894ff8564d7f8edc9R128. Thanks for the suggestion! @msxavi

Reply:
Not sure about your suggestion to use `#update_column` as [just above](https://github.com/plataformatec/devise/pull/5145#discussion_r352292285) I explained why it can't be used. @msxavi

Reply:
Oh sure, no problem. I ended up using `update_column` because `save(validate: false)` didn't work quite well for me due to database constraints. Just to be clear, I'm trying to solve the stretches issue in the product I work on.

That's my final code which seems to work:
```
    return unless self.encrypted_password.present?
    if BCrypt::Password.new(self.encrypted_password).cost != self.class.stretches && persisted?
      update_column(:encrypted_password, BCrypt::Password.create(password, cost: self.class.stretches).to_s)
    end
```

Reply:
Curious why `save(validate: false)` didn't work in your case.  

I use similar code with `#update_column` in a monkey-patched `Devise::Strategies::DatabaseAuthenticatable`, but this PR is to replace this monkey-patched solution.